### ISSUE TITLE
[Patterns] distributed histogramming

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#19102]
This issue is part of a series of issues to design the interface for collections across serial, parallel, and distributed contexts. Our goal is to determine what we think is reasonable to support and what doesn't seem useful. Additional patterns welcome (but it would be best to put them in their own issue for discussion, likely).  In particular, this pattern was suggested by Michael.

```chapel
// suppose hist is a distributed map, input is a distributed array
forall key in input {
  hist[key] += 1;
}
```

This could cause a race depending on the underlying implementation.  Adding a new key in another thread could cause the underlying data structure to resize, and that could make the reference to `hist[key]` in this thread become invalidated.

If we don't return a reference from `hist[key]`, it becomes a lot harder to implement a histogram using this map.  If we don't allow resizing to impact already existing keys, we still could run into reference invalidation if a key is removed in a parallel setting (solutions 1 and 2 below would avoid that case).

Michael had a few ideas for what to do if we don't return a reference from `hist[key]`:

1. We could pass in a lambda/function object to perform the update, something like `Count.update(key, lambda(x) { x += 1; } )`
2. We could use a context manager to wrap the update: `with Count.update(key) as val { val += 1; }`
3. We could add a lock per key and make the user deal with atomicity
4. We could add a feature like the one proposed in #12306

**Comment by bradcray** [comment]
This challenges seems vaguely related to #17999 in that part of the issue is that we're relying on a `ref`-returning interface.  I don't know whether any of the solutions being proposed there might help here.

I think the thing that worries me most about this pattern is that even if we had one of the proposed solutions implemented and working, how do we protect the naive user from shooting themselves in the foot in this way? Oh, or wait... maybe we already have that through task intents:  `hist` would be passed to the tasks by `with (const ref)` which should prevent us from modifying it in this way (a quick check suggests that this is the case).  So that's a bit reassuring...

**Comment by lydia-duncan** [comment]
An extension is considering how each of the potential solutions work when trying to update multiple maps at the same time

**Comment by mppf** [comment]
> This challenges seems vaguely related to #17999 in that part of the issue is that we're relying on a `ref`-returning interface.  I don't know whether any of the solutions being proposed there might help here.

Right, notionally we could have the `map` implement `proc access+=(idx, amt)` and have the compiler arrange for `map[idx] += amt` to call it.

**Comment by mppf** [comment]
I don't see it above, but a possible solution is for our data structures to not do reference invalidation when adding elements. This addresses the main problem here but might limits implementations more than we like. (E.g. ConcurrentMap can't do this, AFAIK).

**Comment by mppf** [comment]
See also #8339 about the way the obvious implementation is a race condition today.
